### PR TITLE
Add MLP model trainer

### PR DESCRIPTION
**Pull request by deerzq**
Initial version of the MLP trainer with simple unit tests. will add more tests in next PR

to: @hectorgon @christhetree 
cc: @jq

### CODE REVIEW

File: training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+      case "memory" => raw.cache()
+      case _ : String => raw
+    }
+
+    val model = setupModel(trainerOptions, networkOptions, pointwise)
+
+    modelIteration(sc, trainerOptions, model, pointwise)
+
+    trainerOptions.cache match {
+      case "memory" => pointwise.unpersist()
```
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala]: case _ => Unit or else it will crash

File: training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+  private def computeActivationGradient(activation: Double,
+                                        func: FunctionForm): Double = {
+    // compute the gradient of activation w.r.t input
+    // note: input only need to be specified for RELU
+    func match {
+      case FunctionForm.SIGMOID => activation * (1.0 - activation)
+      case FunctionForm.RELU => if (activation > 0) 1.0 else 0.0
+      case FunctionForm.IDENTITY => 1.0
+      case FunctionForm.TANH => 1.0 - activation * activation
+      case _ => activation * (1.0 - activation)
```
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala]: I would assert fail since this is a training function

**Comment by deerzq** [comment]
Just added some unit tests.

File: training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+  }
+
+  private def computeUpdates(prevUpdate: FloatVector,
+                             momentum: Float,
+                             learningRate: Float,
+                             gradient: FloatVector): FloatVector = {
+    // based on hinton's dropout paper: http://arxiv.org/pdf/1207.0580.pdf
+    val update: FloatVector = new FloatVector(prevUpdate.length)
+    update.multiplyAdd(momentum, prevUpdate)
+    update.multiplyAdd(-(1.0f - momentum) * learningRate, gradient)
```
Comment (Patrick1860) [inline training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala]: reference to this formula?  I saw one version is mu * v - learning_rate * dx which may transfer to update.multiplyAdd(momentum, prevUpdate)
update.multiplyAdd(-(1.0f ) * learningRate, gradient)

Comment (Patrick1860) [inline training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala]: reference to this formula?  I saw one version is mu * v - learning_rate * dx which transfers to update.multiplyAdd(momentum, prevUpdate)
update.multiplyAdd(-(1.0f ) * learningRate, gradient)

File: training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+    for (i <- (0 until numHiddenLayers).reverse) {
+      // i decreases from numHiddenLayers-1 to 0
+      val numNode = layerNodeNumber.get(i)
+      val numNodeUpperLayer = layerNodeNumber.get(i + 1)
+      val func = activationFunctions.get(i)
+      val thisLayerDelta = new FloatVector(numNode)
+      // compute gradient of weights from the i-th layer to the (i+1)-th layer
+      val activations = model.getLayerActivations.get(i)
+      val hiddenLayerWeights = model.getHiddenLayerWeights.get(i)
+      val biasKey = (LAYER_PREFIX + i.toString, BIAS_PREFIX)
```
Comment (Patrick1860) [inline training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala]: not sure if it is more clear to call it UpperLayerBiasKey as it is the input of this layer

File: training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+import org.apache.spark.SparkContext
+import org.apache.spark.SparkContext._
+import org.apache.spark.rdd.RDD
+
+import scala.collection.JavaConversions._
+import scala.util.Try
+
+/**
+  * A trainer that generates a MLP model.
+  * TODO (Peng): add maxNorm regularizations
```
Comment (Patrick1860) [inline training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala]: great~

File: training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+      val numNodeUpperLayer = layerNodeNumber.get(i + 1)
+      val func = activationFunctions.get(i)
+      val thisLayerDelta = new FloatVector(numNode)
+      // compute gradient of weights from the i-th layer to the (i+1)-th layer
+      val activations = model.getLayerActivations.get(i)
+      val hiddenLayerWeights = model.getHiddenLayerWeights.get(i)
+      val biasKey = (LAYER_PREFIX + i.toString, BIAS_PREFIX)
+      val biasGrad = gradient.getOrElse(biasKey, new FloatVector(numNode))
+      for (j <- 0 until numNode) {
+        val key = (LAYER_PREFIX + i.toString, NODE_PREFIX + j.toString)
```
Comment (Patrick1860) [inline training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala]: add some comments here about how the drop out plays here in the backPropagation process? e.g how the gradient of parameters that related to the omitted note set to 0 in this mini batch?

Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala]: the drop out part is implemented in the model, because the activations corresponding to dropped nodes are zeros, the gradient for the weights that connecting that activation with the next node would be 0, so we don' t need to take special care here. will add some comments

**Comment by deerzq** [comment]
@hectorgon updated the PR to address your comments, PTAL? Thanks!

**Comment by hectorgon** [comment]
:+1:  but did you address Pai's comment about the momentum?

**Comment by deerzq** [comment]
@hectorgon there are different implementations for momentum, I am using the formula in http://arxiv.org/pdf/1207.0580.pdf

File: training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala (hunk 6 above)
Comment (Patrick1860) [inline training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala]: :+1:

**Comment by Patrick1860** [comment]
LGTM didn't see the implementation details of the paper that is hiding after the reference section :).

**Comment by deerzq** [comment]
Thanks for reviewing, will add multi-class part for the trainer and properly test with 20 news.
